src/NuGet.Core/NuGet.Commands/CommandRunners/PackageSearchRunner.cs

```diff
@@ -0,0 +1,198 @@
+        public static async Task RunAsync(
+            IPackageSourceProvider sourceProvider,
+            List<string> sources, string searchTerm,
+            int skip,
+            int take,
+            bool prerelease,
+            bool exactMatch,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var listEndpoints = GetEndpointsAsync(sources, sourceProvider);
+            WarnForHTTPSources(listEndpoints, logger);
+
+            List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)> taskList;
+            if (exactMatch)
+            {
+                taskList = await BuildTaskListForExactMatch(listEndpoints, searchTerm, prerelease, logger, cancellationToken);
+            }
+            else
+            {
+                taskList = await BuildTaskListForSearch(listEndpoints, searchTerm, prerelease, skip, take, logger, cancellationToken);
+
+            }
+
+            var result = new PackageSearchResult(
+                taskList,
+                logger,
+                searchTerm,
+                exactMatch);
+            await result.PrintResultTablesAsync();
+        }
+
+        /// <summary>
+        /// Builds a list of tasks that perform package search operations.
+        /// </summary>
+        /// <param name="listEndpoints">The list of package sources/endpoints.</param>
+        /// <param name="searchTerm">The term to search for within the package sources.</param>
+        /// <param name="prerelease">A flag indicating whether to include prerelease packages in the search.</param>
+        /// <param name="skip">The number of results to skip.</param>
+        /// <param name="take">The number of results to retrieve.</param>
+        /// <param name="logger">The logger instance to use for logging.</param>
+        /// <param name="cancellationToken">A token to monitor for cancellation requests.</param>
+        /// <returns>A list of tasks that perform package search operations.</returns>
+        private static async Task<List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)>> BuildTaskListForSearch(
+            IList<PackageSource> listEndpoints,
+            string searchTerm,
+            bool prerelease,
+            int skip,
+            int take,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var taskList = new List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)>();
+
+            foreach (var source in listEndpoints)
+            {
+                var repository = Repository.Factory.GetCoreV3(source);
+                var resource = await repository.GetResourceAsync<PackageSearchResource>(cancellationToken);
+                taskList.Add(
+                    resource == null ? (null, source) :
+                    (resource.SearchAsync(searchTerm, new SearchFilter(includePrerelease: prerelease), skip, take, logger, cancellationToken), source)
```

Comment:
I'm not a fan of doing too many things in a single statement (I care most about what the debugger will stop at and let me inspect variables, not what the compiler's definition of a statement is). Here, I can break before the `taskList.Add`, but a single "step over" will pass over the `new SearchFilter` and `resource.SearchAsync` results, not to mention the `? :` expression as well.

If you add variables, so that these values are inspectable with the debugger, when compiled in release configuration (with optimization), the variables are optimized away, so there's no performance benefit.